win32/service.py

```diff
@@ -0,0 +1,289 @@
+    devnull = None
+
+    def __init__(self, args):
+        win32serviceutil.ServiceFramework.__init__(self, args)
+
+        AgentSvc.devnull = open(os.devnull, 'w')
+
+        config = get_config(parse_args=False, can_query_registry=False)
+        if config['api_key'] == 'APIKEYHERE':
+            self._update_config_file(config)
+
+        # Let's have an uptime counter
+        self.start_ts = None
+
+        # find the main agent dir, for instance C:\Program Files\Datadog\Datadog Agent\agent
+        dd_dir = self._find_dd_dir()
+        # clean the env vars
+        agent_env = self._prepare_agent_env(dd_dir)
+        # cd to the agent dir to easily launch the 4 components
+        os.chdir(os.path.join(dd_dir, 'agent'))
+
+        embedded_python = os.path.join(dd_dir, 'embedded', 'python.exe')
+        # Keep a list of running processes so we can start/end as needed.
+        # Processes will start started in order and stopped in reverse order.
+        self.procs = {
+            'forwarder': DDProcess(
+                "forwarder",
+                [embedded_python, "ddagent.py"],
+                agent_env
+            ),
+            'collector': DDProcess(
+                "collector",
+                [embedded_python, "agent.py", "foreground", "--use-local-forwarder"],
+                agent_env
+            ),
+            'dogstatsd': DDProcess(
+                "dogstatsd",
+                [embedded_python, "dogstatsd.py", "--use-local-forwarder"],
+                agent_env,
+                enabled=config.get("use_dogstatsd", True)
+            ),
+            'jmxfetch': JMXFetchProcess(
+                "jmxfetch",
+                [embedded_python, "jmxfetch.py"],
+                agent_env,
+                max_restarts=self._MAX_JMXFETCH_RESTARTS,
+                enabled=self._is_jmxfetch_enabled(config)
+            ),
+        }
+
+    def _find_dd_dir(self):
+        # This file is somewhere in the dist directory of the agent
+        file_dir = os.path.dirname(os.path.realpath(__file__))
+        search_dir, current_dir = os.path.split(file_dir)
+        # So we go all the way up to the dist directory to find the actual agent dir
+        while current_dir and current_dir != 'dist':
+            search_dir, current_dir = os.path.split(search_dir)
+        dd_dir = search_dir
+        # If we don't find it, we use the default
+        if not current_dir:
+            dd_dir = os.path.join('C:\\', 'Program Files', 'Datadog', 'Datadog Agent')
+
+        return dd_dir
+
+    def _prepare_agent_env(self, dd_dir):
+        # preparing a clean env for the agent processes
+        env = os.environ.copy()
+        if env.get('PYTHONPATH'):
+            del env['PYTHONPATH']
+        if env.get('PYTHONHOME'):
+            del env['PYTHONHOME']
+        if env['PATH'][-1] != ';':
+            env['PATH'] += ';'
+        env['PATH'] += "{};{};".format(os.path.join(dd_dir, 'bin'), os.path.join(dd_dir, 'embedded'))
```

Comment:
Because `gohai` requires `python` to be in the path (to get the python version), and `python.exe` is in `embedded`.